File: apps/web/test/category-order.test.ts
import { describe, expect, it } from 'vitest';
import { readCategories } from '@/app/data/categories';
import { FakeServer } from './fake-server';
import { device } from './helpers';

/**
 * Custom categories are listed by when each was made, as the phone
 * lists them. A restore bumps `updatedAt` so the undo syncs, and must
 * not move the category to the end of the list for it.
 */
describe('the order of custom categories', () => {
  it('keeps a deleted and restored category where it was', async () => {
    const h = await device(new FakeServer());
    const books = await h.categories.create('Books', 'school');
    h.clock.advance(1000);
    await h.categories.create('Coffee', 'coffee');
    h.clock.advance(1000);
    await h.categories.create('Gifts', 'gift');
    h.clock.advance(1000);

    await h.categories.remove(books);
    h.clock.advance(1000);
    await h.categories.restore(books);

    const rows = await readCategories(h.db);
    expect(rows.map((row) => row.name)).toEqual(['Books', 'Coffee', 'Gifts']);
    // The restore is still an edit the other devices hear about.
    const restored = rows[0]!;
    expect(restored.createdAt).toBe('2026-09-19T12:00:00.000Z');
    expect(restored.updatedAt).toBe('2026-09-19T12:00:04.000Z');
  });

  it('orders a category from before createdAt by its last edit', async () => {
    const h = await device(new FakeServer());
    h.clock.advance(60_000);
    await h.categories.create('Coffee', 'coffee');
    await h.db.rows('expense_categories').put({
      uuid: 'old',
      name: 'Books',
      icon: 'school',
      createdAt: null,
      deletedAt: null,
      updatedAt: '2026-09-19T12:00:30.000Z',
    });

    expect((await readCategories(h.db)).map((row) => row.name)).toEqual(['Books', 'Coffee']);
  });
});

File: apps/mobile/test/features/finances/category_order_test.dart
import 'package:drift/drift.dart' hide isNotNull, isNull;
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:harvest/core/db/database.dart';
import 'package:harvest/features/finances/data/finances_repository.dart';

/// Custom categories are listed by when each was made. A restore (from
/// the web, arriving through sync) bumps `updated_at`, and must not move
/// the category to the end of the list for it.
void main() {
  late HarvestDatabase db;
  late FinancesRepository repo;

  setUp(() {
    db = HarvestDatabase.forTesting(NativeDatabase.memory());
    repo = FinancesRepository(db);
  });

  tearDown(() => db.close());

  Future<List<String>> names() async =>
      (await repo.watchCategories().first).map((c) => c.name).toList();

  test('a new category remembers when it was made', () async {
    await repo.createCategory(name: 'Books', icon: 'school');
    final row = await db.select(db.expenseCategories).getSingle();
    expect(row.createdAt, isNotNull);
  });

  test('a restored category keeps its place', () async {
    final made = DateTime.utc(2026, 9);
    for (final (i, name) in ['Books', 'Coffee', 'Gifts'].indexed) {
      final at = made.add(Duration(hours: i));
      await db
          .into(db.expenseCategories)
          .insert(
            ExpenseCategoriesCompanion.insert(
              uuid: name,
              name: name,
              icon: 'category',
              createdAt: Value(at),
              updatedAt: Value(at),
            ),
          );
    }
    await repo.deleteCategory('Books');
    await (db.update(
      db.expenseCategories,
    )..where((c) => c.uuid.equals('Books'))).write(
      ExpenseCategoriesCompanion(
        deletedAt: const Value(null),
        updatedAt: Value(DateTime.utc(2026, 9, 20)),
      ),
    );

    expect(await names(), ['Books', 'Coffee', 'Gifts']);
  });

  test('a category from before created_at orders by its last edit', () async {
    await db
        .into(db.expenseCategories)
        .insert(
          ExpenseCategoriesCompanion.insert(
            uuid: 'coffee',
            name: 'Coffee',
            icon: 'coffee',
            createdAt: Value(DateTime.utc(2026, 9, 2)),
            updatedAt: Value(DateTime.utc(2026, 9, 2)),
          ),
        );
    await db
        .into(db.expenseCategories)
        .insert(
          ExpenseCategoriesCompanion.insert(
            uuid: 'books',
            name: 'Books',
            icon: 'school',
            updatedAt: Value(DateTime.utc(2026, 9)),
          ),
        );

    expect(await names(), ['Books', 'Coffee']);
  });
}
